### ISSUE TITLE
Macro Revamp

### ISSUE THREAD

**Issue by mgattozzi** [mgattozzi/github-rs#114]
Currently the way macros work is kind of unweildy. I have started a separate library based of github-rs that can be reused here eventually but it's not ready for use just yet. Most of the current code will have to be rewritten but everything should be easier to maintain overall.

**Comment by jaroslaw-weber** [comment]
do you think it is possible to generate whole api from api list?
for example
GET /users/:username/repos
would generate part for getting user repos.

if it is difficult to do it with macros, I think it may be possible with some layer of transpiling.

it would allow to just copy paste endpoints to implement them.

**Comment by mgattozzi** [comment]
@jaroslaw-weber this would be really cool. I spun out a project [rust-web-apis/foundation](https://github.com/rust-web-apis/foundation/) to kind of do this at some point making the macros here easier to deal with. Ultimately I think it'll come down to procedural macros so that some compile time code can be run. The macro rules here work but they're limited. You can't put in a string with :user as a url param and expect the macro to know what that means. I think we could end up redoing this all with proc_macro_derive but use it for proc_macro and still keep things on stable as a result.

**Comment by jaroslaw-weber** [comment]
Great project idea! Making it easier to create a wrapper for REST apis is the way to go!
I tried very simple code generation approach (but without macros):
https://github.com/jaroslaw-weber/github_autogen

**Comment by mgattozzi** [comment]
@jaroslaw-weber Yeah I built a lot of these macros as I was learning them and well they're not the easiest to contribute too. I'd like to clean them all up at some point. Having other things to do and other projects makes it a bit hard to do.
